pymdownx/snippets.py

```diff
@@ -190,43 +195,65 @@ def download(self, url):
-                # content-length was in the header, so we did not read yet
-                content = response.read()
-
-            # Process lines
-            last = content.endswith((b'\r', b'\n'))
-            s_lines = [l.decode(self.encoding) for l in content.splitlines()]
-            if last:
-                s_lines.append('')
-            return s_lines
+        for attempt in range(1, self.max_retries + 1):
+            try:
+                http_request = urllib.request.Request(url, headers=self.url_request_headers)
+                timeout = None if self.url_timeout == 0 else self.url_timeout
+                with urllib.request.urlopen(http_request, timeout=timeout) as response:
+                    # Fail if status is not OK
+                    status = response.status if util.PY39 else response.code
+                    
+                    if status == 429:  # Rate limited
+                        if attempt < self.max_retries:
+                            wait = self.backoff_factor * attempt
+                            time.sleep(wait)
+                            continue
+                        else:
+                            raise SnippetMissingError(f"Rate limited after {self.max_retries} attempts")
+
+                    if status != 200:
+                        raise SnippetMissingError(f"Cannot download snippet '{url}'")
+
+                    # We provide some basic protection against absurdly large files.
+                    # 32MB is chosen as an arbitrary upper limit. This can be raised if desired.
+                    content = None
+                    if "content-length" not in response.headers:
+                        # we have to read to know if we went over the max, but never more than `url_max_size`
+                        # where `url_max_size` == 0 means unlimited
+                        content = response.read(self.url_max_size) if self.url_max_size != 0 else response.read()
+                        content_length = len(content)
+                    else:
+                        content_length = int(response.headers["content-length"])
+
+                    if self.url_max_size != 0 and content_length >= self.url_max_size:
+                        raise ValueError(f"refusing to read payloads larger than or equal to {self.url_max_size}")
+
+                    # Nothing to return
+                    if content_length == 0:
+                        return ['']
+
+                    if content is None:
+                        # content-length was in the header, so we did not read yet
+                        content = response.read()
+
+                    # Process lines
+                    last = content.endswith((b'\r', b'\n'))
+                    s_lines = [l.decode(self.encoding) for l in content.splitlines()]
+                    if last:
+                        s_lines.append('')
+                    return s_lines
+
+            except urllib.error.HTTPError as e:
```

Comment:
I thought we were checking status earlier, why would we now need to check it here? Are we actually getting a raised exception now instead of just being able to check the status? This isn't inherently bad per se, but I'm curious what's actually getting exercised in the code.

Reply:
I do see in the urlib code that some status can raise an exception. It looks like 429 may do this. If that is the case, then the earlier 429 check is not needed.